api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/AdvancedDiseaseAndTBCascadeCohortQueries.java

```diff
@@ -0,0 +1,1485 @@
+    cd.addParameter(new Parameter("endDate", "End Date", Date.class));
+    cd.addParameter(new Parameter("startDate", "Start Date", Date.class));
+
+    CohortDefinition absoluteCd4 = getPatientsWithAbsoluteCd4Count(cd4);
+    CohortDefinition age = ageCohortQueries.createXtoYAgeCohort("Age", minAge, maxAge);
+
+    cd.addSearch("absoluteCd4", EptsReportUtils.map(absoluteCd4, mappings));
+
+    cd.addSearch("age", EptsReportUtils.map(age, "effectiveDate=${endDate}"));
+
+    cd.setCompositionString("absoluteCd4 AND age");
+
+    return cd;
+  }
+
+  /**
+   * <b>The system will include all clients who reinitiated ART during the inclusion period who
+   * have:</b>
+   *
+   * <ul>
+   *   <li>The most recent “Mudança de Estado de Permanência” state marked as “Reinício” in Ficha
+   *       Clínica, with the state date falling during the inclusion period.
+   *   <li>OR The most recent “Mudança de Estado de Permanência” state marked as “Reinício” in Ficha
+   *       Resumo, with the state date falling during the inclusion period.
+   * </ul>
+   *
+   * @return CohortDefinition *
+   */
+  public CohortDefinition getPatientsWhoReinitiatedArt() {
+    SqlCohortDefinition cd = new SqlCohortDefinition();
+    cd.setName("Clients who reinitiated ART during the inclusion period");
+    cd.addParameter(new Parameter("startDate", "Start Date", Date.class));
+    cd.addParameter(new Parameter("endDate", "End Date", Date.class));
+    cd.addParameter(new Parameter("location", "End Date", Location.class));
+
+    String query =
+        "SELECT p.patient_id "
+            + "FROM   patient p "
+            + "       INNER JOIN (SELECT e.patient_id "
+            + "                   FROM   encounter e INNER JOIN obs o ON o.encounter_id = e.encounter_id "
+            + "                          INNER JOIN(SELECT e.patient_id, "
+            + "                                            Max(e.encounter_datetime) state_date "
+            + "                                     FROM   encounter e INNER JOIN obs o ON o.encounter_id = e.encounter_id "
+            + "                                     WHERE  e.voided = 0 "
+            + "                                            AND o.voided = 0 "
+            + "                                            AND e.encounter_type = ${6} "
+            + "                                            AND o.concept_id = ${6273} "
+            + "                                            AND e.location_id = :location "
+            + "                                            AND e.encounter_datetime BETWEEN :startDate AND :endDate "
+            + "                                     GROUP  BY e.patient_id) recent_state "
+            + "                                  ON recent_state.patient_id = e.patient_id "
+            + "                   WHERE  e.voided = 0 "
+            + "                          AND o.voided = 0 "
+            + "                          AND o.concept_id = ${6273} "
```

Comment:
@eliazmuchave  I think here is missing the encounter_type

Reply:
Fixed